### ISSUE TITLE
Highlight diffs when there is unrecognised initial content

### ISSUE THREAD

**Issue by dandavison** [dandavison/delta#121]
cc @m-lima @gibfahn Currently, if delta doesn't recognize the type of diff from the first line, it [gives up](https://github.com/dandavison/delta/blob/master/src/delta.rs#L72) on parsing. I'm aware of two situations where this is too strict:

1. [Magit](https://github.com/magit/magit) makes calls like `git show -p --cc --format=  --numstat --stat`. This results in the `numstat` and `stat` output appearing before the diff, and the diff is not highlighted as a result. (I'm making it possible to use `delta` as a diff rendering plugin for magit.)

2. In #100 delta does not render `git stash show --stat --patch` correctly for the same reason.

**Comment by m-lima** [comment]
I suppose the crux here is to extend [`detect_source`](https://github.com/dandavison/delta/blob/master/src/delta.rs#L171) to detect a source based on multiple lines (without consuming them)?

**Comment by dandavison** [comment]
> I suppose the crux here is to extend detect_source to detect a source based on multiple lines (without consuming them)?

I was hoping to be able to more flexible and not care what the initial content is until we get to recognizable content. If you would like to, please feel free to comment on https://github.com/dandavison/delta/pull/123!

**Comment by gibfahn** [comment]
>I was hoping to be able to more flexible and not care what the initial content is until we get to recognizable content. If you would like to, any thoughts you have on #123 would be appreciated!

Yeah this sounds really useful, it means we could use delta in places where there is arbitrary text before the diff.

Would this also handle unrecognised text after the diff as well, or is that not possible?

**Comment by dandavison** [comment]
> Would this also handle unrecognised text after the diff as well, or is that not possible?

Yes, unrecognised text after the diff is printed out. However, currently, the state machine will still be in HUNK mode (see screenshot below, in particular the way that the initial junk is not syntax highlighted but when it occurs further down after a hunk it is.) This can be improved: basically, when in a diff hunk, the first character in each line must be one of `{'+', '-', ' '}`. However, the current code in #123 doesn't enforce that (I'll put a comment in the code to show why). Thanks, that was a helpful question -- the behaviour below  should be fixed either in #123 or perhaps in a subsequent PR.

```
(
    echo xxxxxxx
    echo "((;;);$~(()"
    git diff --stat --numstat --patch
    echo xxxxxxx
    echo "((;;);$~(()"
    git diff --stat --numstat --patch
    echo xxxxxxx
    echo "((;;);$~(()"
) | delta
```

<table><tr><td><img width=300px src="https://user-images.githubusercontent.com/52205/79072824-2466e480-7cb1-11ea-974b-2027f4e58824.png" alt="image" /></td></tr></table>
